modules/mod-deep-learning/DeepModelManager.h

```diff
@@ -0,0 +1,131 @@
+/**********************************************************************
+
+   Audacity: A Digital Audio Editor
+   Audacity(R) is copyright (c) 1999-2021 Audacity Team.
+
+   DeepModelManager.h
+   Hugo Flores Garcia
+
+******************************************************************/
+/**
+
+\file DeepModelManager.h
+\brief tools for downloading and managing Audacity models hosted in HuggingFace
+
+*/
+/*******************************************************************/
+
+#pragma once
+
+#include "DeepModel.h"
+#include "ModelCard.h"
+
+#include "Request.h"
+#include "NetworkManager.h"
+#include "widgets/ProgressDialog.h"
+#include "AudacityException.h"
+
+// this exception should be caught internally, but we 
+// derive from MessageBoxException just in case it needs to 
+// get handled by Audacity
+class ModelManagerException final : public MessageBoxException
+{
+public:
+   ModelManagerException(const TranslatableString msg) : 
+                        m_msg(std::move(msg)),
+                        MessageBoxException{
+                           ExceptionType::Internal,
+                           XO("Model Manager Error")
+                        } 
+   {}
+
+   //! internal message
+   virtual const char* what() const noexcept
+      { return m_msg.Translation().c_str(); }
+
+   //! user facing message
+   virtual TranslatableString ErrorMessage() const
+      { return XO("Model Manager Error: \n %s").Format(m_msg);}
+
+   const TranslatableString m_msg;
+};
+
+using RepoIDList = std::vector<std::string>;
+
+// callbacks
+using CompletionHandler = std::function<void (int httpCode, std::string responseBody)>;
+using ProgressCallback = std::function<void(int64_t current, int64_t expected)>;
+
+using ModelSizeCallback = std::function<void(size_t size)>;
+using CardFetchProgressCallback = std::function<void(int64_t current, int64_t expected)>;
+using RepoListFetchedCallback = std::function<void(bool success, RepoIDList repos)>;
+using CardFetchedCallback = std::function<void(bool succcess, ModelCardHolder card)>;
+
+class DeepModelManager final
+{
+   // private! Use Get()
+   DeepModelManager(); 
+   ~DeepModelManager() = default;
```

Comment:
When making a "singleton" class like this, these lines, to prohibit copying, are also useful:
```
DeepModelManager(const & DeepModelManager) = delete;
DeepModelManager &operator=(const & DeepModelManager) = delete;
```
The are redundant to stop copying in this case, because of the private destructor and the `std::mutex` member, but still it makes the intent plain and the impossibility of copy more obvious to the reviewer.

Banning copy causes the compiler to flag this mistake:
```
auto manager = DeepModelManager::Get();
```
where `&` to declare a reference variable was omitted.

I have been guilty of such mistakes.

Reply:
30f12ed7fc19470527419f8782fceb3ca7c15386